doc/tapscript_opcodes.md

```diff
@@ -0,0 +1,96 @@
+   - Define `OP_SUCCESS197` as `OP_SHA256UPDATE` which pops 1) a SHA256 context and 2) bytestring and push an updated context by adding the bytestring to the data stream being hashed.
+   - Define `OP_SUCCESS198` as `OP_SHA256FINALIZE` to pop a SHA256 context and bytestring and push a SHA256 hash value after adding the bytestring and completing the padding.
+
+
+2. **Transaction Introspection codes**: Transaction introspection is already possible in elements script by use of `OP_CHECKSIGFROMSTACKVERIFY`, however the current solutions are really expensive in applications like [covenants](https://github.com/sanket1729/covenants-demo). Therefore, we are not adding any new functionality by supporting introspection, only making it easier to use. The warning still remains the same as with covenants, if the user is inspecting data from parts of the transaction that are not signed, the script can cause unexpected behavior.
+For opcodes that inspect data that is not committed in sighash, introspection is safe because any changes to the witness data would cause wtxid to change and it would revalidate the tx again.
+   - Define `OP_SUCCESS199` as `OP_INSPECTINPUT` with the following semantics: Pop two minimal `CScriptNum`s from stack. The first number `n` denotes the type of introspection, and the second number `idx `denotes the position of input to inspect. Immediately abort if `idx` is out of bounds.
+      1. If `n=0`, `OP_INSPECTINPUTSPENDTYPE` Push 1 byte "spend_type" onto the stack. spend_type (1) is equal to `(ext_flag * 2) + annex_present` as defined in [Modified BIP-341 SigMsg for Elements](https://gist.github.com/roconnor-blockstream/9f0753711153de2e254f7e54314f7169)
+      2. If `n=1`, `OP_INSPECTINPUTOUTPOINTFLAG` Push the outpoint_flag(1) as defined in [Modified BIP-341 SigMsg for Elements](https://gist.github.com/roconnor-blockstream/9f0753711153de2e254f7e54314f7169)
+      3. if `n=2`, `OP_INSPECTINPUTOUTPOINT` Push the outpoint as a tuple. First push the `txid`(32) of the `prev_out`, followed by a 4 byte push of `vout`
+      4. If `n=3`, `OP_INSPECTINPUTASSET` Push the `nAsset` onto the stack as two elements. The first push the assetID(32), followed by the prefix(1)
+      5. If `n=4`, `OP_INSPECTINPUTVALUE` Push the `nValue` as a tuple, value(8, 32) followed by prefix,
+      6. If `n=5`, `OP_INSPECTINPUTSSCRIPTPUBKEY` Push the scriptPubkey(35) onto the stack.
+      7. If `n=6`, `OP_INSPECTINPUTSEQUENCE` Push the `nSequence`(4) as little-endian number.
+      8. If `n=7`, `OP_INSPECTINPUTISSUANCE` Push the assetIssuance information(74-130) if the asset has issuance, otherwise push an empty vector
+      9. If `n=8`, `OP_INSPECTINPUTANNEX` Push the annex onto the stack where the annex includes the prefix(0x50). If the annex does not exist, push an empty vector
+      10. Otherwise fail.
```

Comment:
Maybe we shoud "otherwise end the script with success" so that we can softfork in new data?

Reply:
It is somewhat risky. If the user does some manipulation on the stack to figure out which value `n` to use and some makes an error in it, it could result in loss of funds. 

I guess the same could be said for `k` in `k on n` multisig. So, I feel safe in changing it to success.